drivers/flash/flash_stm32l5x.c

```diff
@@ -0,0 +1,381 @@
+	uint32_t tmp;
+	int rc;
+	int page;
+
+	/* if the non-secure control register is locked,do not fail silently */
+	if (regs->NSCR & FLASH_NSCR_NSLOCK) {
+		LOG_ERR("NSCR locked\n");
+		return -EIO;
+	}
+
+	/* Check that no Flash memory operation is ongoing */
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	if ((regs->OPTR & FLASH_OPTR_DBANK) == FLASH_OPTR_DBANK) {
+		bool bank_swap;
+		/* Check whether bank1/2 are swapped */
+		bank_swap =
+		((regs->OPTR & FLASH_OPTR_SWAP_BANK) == FLASH_OPTR_SWAP_BANK);
+
+		if ((offset < (FLASH_SIZE / 2)) && !bank_swap) {
+			/* The pages to be erased is in bank 1 */
+			regs->NSCR &= ~FLASH_NSCR_NSBKER_Msk;
+			page = offset / FLASH_PAGE_SIZE;
+			LOG_DBG("Erase page %d on bank 1", page);
+		} else if ((offset >= BANK2_OFFSET) && bank_swap) {
+			/* The pages to be erased is in bank 1 */
+			regs->NSCR &= ~FLASH_NSCR_NSBKER_Msk;
+			page = (offset - BANK2_OFFSET) / FLASH_PAGE_SIZE;
+			LOG_DBG("Erase page %d on bank 1", page);
+		} else if ((offset < (FLASH_SIZE / 2)) && bank_swap) {
+			/* The pages to be erased is in bank 2 */
+			regs->NSCR |= FLASH_NSCR_NSBKER;
+			page = offset / FLASH_PAGE_SIZE;
+			LOG_DBG("Erase page %d on bank 2", page);
+		} else if ((offset >= BANK2_OFFSET) && !bank_swap) {
+			/* The pages to be erased is in bank 2 */
+			regs->NSCR |= FLASH_NSCR_NSBKER;
+			page = (offset - BANK2_OFFSET) / FLASH_PAGE_SIZE;
+			LOG_DBG("Erase page %d on bank 2", page);
+		} else {
+			LOG_ERR("Offset %d does not exist", offset);
+			return -EINVAL;
+		}
+	} else {
+		page = offset / FLASH_PAGE_SIZE;
+		LOG_DBG("Erase page %d\n", page);
+	}
+
+	/* Set the NSPER bit and select the page you wish to erase */
+	regs->NSCR |= FLASH_NSCR_NSPER;
+	regs->NSCR &= ~FLASH_NSCR_NSPNB_Msk;
+	regs->NSCR |= (page << FLASH_NSCR_NSPNB_Pos);
+
+	/* Set the NSSTRT bit */
+	regs->NSCR |= FLASH_NSCR_NSSTRT;
+
+	/* flush the register write */
+	tmp = regs->NSCR;
+
+	/* Wait for the NSBSY bit */
+	rc = flash_stm32_wait_flash_idle(dev);
+
+	if ((regs->OPTR & FLASH_OPTR_DBANK) == FLASH_OPTR_DBANK) {
+		regs->NSCR &= ~(FLASH_NSCR_NSPER | FLASH_NSCR_NSBKER);
+	} else {
+		regs->NSCR &= ~(FLASH_NSCR_NSPER);
+	}
+
+	return rc;
+}
+
+int flash_stm32_block_erase_loop(const struct device *dev,
+				 unsigned int offset,
+				 unsigned int len)
+{
+	unsigned int address = offset;
+	int rc = 0;
+	bool icache_enabled = LL_ICACHE_IsEnabled();
+
+	if (icache_enabled) {
+		/* Disable icache, this will start the invalidation procedure.
+		 * All changes(erase/write) to flash memory should happen when
+		 * i-cache is disabled. A write to flash performed without
+		 * disabling i-cache will set ERRF error flag in SR register.
+		 */
+		rc = stm32l5x_icache_disable();
+		if (rc != 0) {
+			return rc;
+		}
+	}
+
+	for (; address <= offset + len - 1 ; address += FLASH_PAGE_SIZE) {
+		rc = erase_page(dev, address);
+		if (rc < 0) {
+			break;
+		}
+	}
+
+	if (icache_enabled) {
```

Comment:
You can move the check for end of invalidation procedure to stm32l5x_icache_enable()
thus it will avoid code duplication.
Also no need to enable icache in case of error returned by icache_wait_for_invalidate_complete()

Reply:
> Also no need to enable icache in case of error returned by icache_wait_for_invalidate_complete()

If i do not enable i-cache incase of error, it will get disabled for ever. In subsequent flash writes or erase, i am checking if i-cache is enabled.

Reply:
Ok, Right.